File: backend/app/services/user_service.py
from prisma import Prisma

from app.schemas.user_schema import UserCreate, UserUpdate
from app.utils.security import hash_password


async def get_user_by_id(db: Prisma, user_id: str):
    return await db.user.find_unique(where={"id": user_id})


async def get_user_by_email(db: Prisma, email: str):
    return await db.user.find_unique(where={"email": email})


async def create_user(db: Prisma, data: UserCreate):
    return await db.user.create(
        data={
            "email": data.email,
            "password": hash_password(data.password),
            "name": data.name,
        }
    )


async def update_user(db: Prisma, user_id: str, data: UserUpdate):
    update_data = {k: v for k, v in data.model_dump().items() if v is not None}
    return await db.user.update(where={"id": user_id}, data=update_data)


async def deactivate_user(db: Prisma, user_id: str):
    return await db.user.update(where={"id": user_id}, data={"isActive": False})

File: backend/src/services/user.service.ts
import { PrismaClient } from "@prisma/client";
import { hashPassword, verifyPassword } from "../utils/password";

export interface CreateUserInput {
  email: string;
  password: string;
  name?: string;
}

export interface UpdateUserInput {
  name?: string;
  email?: string;
}

export async function getUserById(prisma: PrismaClient, userId: string) {
  return prisma.user.findUnique({ where: { id: userId } });
}

export async function getUserByEmail(prisma: PrismaClient, email: string) {
  return prisma.user.findUnique({ where: { email } });
}

export async function createUser(prisma: PrismaClient, data: CreateUserInput) {
  return prisma.user.create({
    data: {
      email: data.email,
      password: await hashPassword(data.password),
      name: data.name,
    },
  });
}

export async function updateUser(
  prisma: PrismaClient,
  userId: string,
  data: UpdateUserInput
) {
  const updateData = Object.fromEntries(
    Object.entries(data).filter(([, v]) => v !== undefined)
  );
  return prisma.user.update({ where: { id: userId }, data: updateData });
}

export async function deactivateUser(prisma: PrismaClient, userId: string) {
  return prisma.user.update({ where: { id: userId }, data: { isActive: false } });
}

export async function changePassword(
  prisma: PrismaClient,
  userId: string,
  currentPassword: string,
  newPassword: string
): Promise<void> {
  const user = await prisma.user.findUnique({ where: { id: userId } });
  if (!user) throw Object.assign(new Error("Usuário não encontrado."), { statusCode: 404 });

  const valid = await verifyPassword(currentPassword, user.password);
  if (!valid) throw Object.assign(new Error("Senha atual incorreta."), { statusCode: 400 });

  const hashed = await hashPassword(newPassword);
  await prisma.user.update({ where: { id: userId }, data: { password: hashed } });
}
